x-pack/plugins/endpoint/public/embeddables/resolver/types.ts

```diff
@@ -130,13 +130,57 @@ export type CameraState = {
     }
 );
 
+/**
+ * This represents all the raw data (sans statistics, metadata, etc.)
+ * about a particular subject's related events
+ */
+export interface RelatedEventDataEntry {
+  relatedEvents: Array<{
+    relatedEvent: ResolverEvent;
+    relatedEventType: string;
+  }>;
+}
+
+/**
+ * Represents the status of the request for related event data, which will be either the data,
+ * a value indicating that it's still waiting for the data or an Error indicating the data can't be retrieved as expected
+ */
+export type RelatedEventDataResults = RelatedEventDataEntry | 'waitingForRelatedEventData' | Error;
+
+/**
+ * This represents the raw related events data enhanced with statistics
+ * (e.g. counts of items grouped by their related event types)
+ */
+export type RelatedEventDataEntryWithStats = RelatedEventDataEntry & {
+  stats: Record<string, number>;
+};
+
+/**
+ * The status or value of any particular event's related events w.r.t. their valence to the current view.
+ * One of:
+ * `RelatedEventDataEntryWithStats` when results have been received and processed and are ready to display
+ * `waitingForRelatedEventData` when related events have been requested but have not yet matriculated
+ * `Error` when the request for any event encounters an error during service
+ */
+export type RelatedEventEntryWithStatsOrWaiting =
+  | RelatedEventDataEntryWithStats
+  | `waitingForRelatedEventData`
+  | Error;
+
+/**
+ * This represents a Map that will return either a `RelatedEventDataEntryWithStats`
+ * or a `waitingForRelatedEventData` symbol when referenced with a unique event.
+ */
+export type RelatedEventData = Map<ResolverEvent, RelatedEventEntryWithStatsOrWaiting>;
+
 /**
  * State for `data` reducer which handles receiving Resolver data from the backend.
  */
 export interface DataState {
   readonly results: readonly ResolverEvent[];
   isLoading: boolean;
   hasError: boolean;
+  resultsEnrichedWithRelatedEventInfo?: Map<ResolverEvent, RelatedEventDataResults>;
```

Comment:
To discourage anything downstream from growing a dependency on top of it, since @jonathan-buttner PR that obviates this contraption is currently under review.

Reply:
Since `resultsEnrichedWithRelatedEventInfo` is optional, all the code that deals w/ it checks for its presence. When reading the code, I have to consider what would happen if the value is missing. When writing comprehensive tests, we would also have to account for scenarios where the value is missing.

However, the implementation always provides the value. If some code is written like:

```
if (resultsEnrichedWithRelatedEventInfo) {
} else {
  // imagine code here
}
```
Then any code in the `else` block is in fact dead code, as `resultsEnrichedWithRelatedEventInfo` isn't really optional. In the future, developers reading and maintaining this code will have to figure out that all these checks are pointless. They will have to figure out that the value isn't really optional. They may be confused, and think that the model is more complex than it is. They may assume that this value is optional, and that all code needs to deal w/ that fact. 

I think for these reasons, you should mark the field as non-optional.